notify/sns/sns.go

```diff
@@ -81,69 +93,85 @@ func (n *Notifier) Notify(ctx context.Context, alert ...*types.Alert) (bool, err
+				reasonErr := notify.NewErrorWithReason(notify.GetFailureReasonFromStatusCode(statusCode), checkErr)
+				return retryable, reasonErr
+			}
 		}
+		// Fallback for non-API errors or if status code extraction fails.
 		return true, err
 	}
 
-	n.logger.Debug("SNS message successfully published", "message_id", publishOutput.MessageId, "sequence number", publishOutput.SequenceNumber)
+	n.logger.Debug("SNS message successfully published", "message_id", aws.ToString(publishOutput.MessageId), "sequence_number", aws.ToString(publishOutput.SequenceNumber))
 
 	return false, nil
 }
 
-func (n *Notifier) createSNSClient(tmpl func(string) string, tmplErr *error) (*sns.SNS, error) {
-	var creds *credentials.Credentials
-	// If there are provided sigV4 credentials we want to use those to create a session.
-	if n.conf.Sigv4.AccessKey != "" && n.conf.Sigv4.SecretKey != "" {
-		creds = credentials.NewStaticCredentials(n.conf.Sigv4.AccessKey, string(n.conf.Sigv4.SecretKey), "")
-	}
-	sess, err := session.NewSessionWithOptions(session.Options{
-		Config: aws.Config{
-			Region:   aws.String(n.conf.Sigv4.Region),
-			Endpoint: aws.String(tmpl(n.conf.APIUrl)),
-		},
-		Profile: n.conf.Sigv4.Profile,
-	})
-	if err != nil {
-		return nil, err
+func (n *Notifier) createSNSClient(ctx context.Context, tmpl func(string) string, tmplErr *error) (*sns.Client, error) {
+	// Base configuration options that apply to both STS (if used) and the final SNS client.
+	baseCfgOpts := []func(*awsconfig.LoadOptions) error{
+		awsconfig.WithHTTPClient(n.client),
+		awsconfig.WithRegion(n.conf.Sigv4.Region),
 	}
-	if *tmplErr != nil {
-		return nil, notify.NewErrorWithReason(notify.ClientErrorReason, fmt.Errorf("execute 'api_url' template: %w", *tmplErr))
+	if n.conf.Sigv4.Profile != "" {
+		baseCfgOpts = append(baseCfgOpts, awsconfig.WithSharedConfigProfile(n.conf.Sigv4.Profile))
 	}
+	if n.conf.Sigv4.AccessKey != "" {
+		creds := credentials.NewStaticCredentialsProvider(n.conf.Sigv4.AccessKey, string(n.conf.Sigv4.SecretKey), "")
+		baseCfgOpts = append(baseCfgOpts, awsconfig.WithCredentialsProvider(creds))
+	}
+
+	// Final configuration options for the SNS client.
+	snsCfgOpts := baseCfgOpts
 
+	// If a RoleARN is provided, create an STS client to assume the role.
+	// This uses a separate config load to ensure the STS client does not use a custom SNS endpoint.
 	if n.conf.Sigv4.RoleARN != "" {
-		var stsSess *session.Session
-		if n.conf.APIUrl == "" {
-			stsSess = sess
-		} else {
-			// If we have set the API URL we need to create a new session to get the STS Credentials.
-			stsSess, err = session.NewSessionWithOptions(session.Options{
-				Config: aws.Config{
-					Region:      aws.String(n.conf.Sigv4.Region),
-					Credentials: creds,
-				},
-				Profile: n.conf.Sigv4.Profile,
-			})
-			if err != nil {
-				return nil, err
-			}
+		stsCfg, err := awsconfig.LoadDefaultConfig(ctx, baseCfgOpts...)
```

Comment:
This is needed later so it might make sense to call this once and only swap the credentials in this branch.

Reply:
The two calls are required because they configure two different clients:

1. The first LoadDefaultConfig is used to build the STS client. Its specific purpose is to create the AssumeRole provider which will fetch the temporary credentials.

2. The second LoadDefaultConfig builds the final configuration for the SNS client. This step is essential to combine the base options with both the new role provider and the custom apiURL.

We can't just modify the first config (stsCfg) because it's targeted at the STS service and doesn't include the SNS-specific endpoint. This two-step process ensures all options are correctly applied for the final client. Please see line 141,142.